include/glow/Optimizer/GraphOptimizer/PassManager.h

```diff
@@ -0,0 +1,93 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_OPTIMIZER_PASSMANAGER_H
+#define GLOW_OPTIMIZER_PASSMANAGER_H
+
+#include "glow/Optimizer/GraphOptimizer/CompilationContext.h"
+#include "glow/Optimizer/GraphOptimizer/FunctionPass.h"
+#include "glow/Optimizer/GraphOptimizer/FunctionPasses.h"
+
+#include "llvm/ADT/SmallVector.h"
+
+namespace glow {
+
+/// Specifies convergence mode for a FunctionPass.
+enum class PassConvergence {
+  OnePass,         // Run a single pass over the Function.
+  UntilFixedPoint, // Run the pass over the Function until a fixed point is
+                   // reached. Runs DCE between each pass.
+};
+
+/// Specifies a configuration for running a FunctionPass when used in a
+/// FunctionPassPipeline.
+struct FunctionPassConfig {
+  /// ID of the FunctionPass to run.
+  FunctionPassID passID{FunctionPassID::EmptyPass};
+
+  /// Convergence mode to inform the PassManager how to run the FunctionPass.
+  PassConvergence convergence{PassConvergence::OnePass};
+
+  /// Which modes the FunctionPass should be run in.
+  llvm::SmallSet<CompilationMode, 2> enabledModes;
+
+  FunctionPassConfig(FunctionPassID ID_,
+                     PassConvergence convergence_ = PassConvergence::OnePass,
+                     std::set<CompilationMode> enabledModes_ =
+                         {CompilationMode::Infer, CompilationMode::Train})
+      : passID(ID_), convergence(convergence_) {
+    for (auto &mode : enabledModes_) {
+      enabledModes.insert(mode);
+    }
+  }
+};
+
+using FunctionPassPipeline = llvm::SmallVector<FunctionPassConfig, 32>;
+
+/// Manager for running a series of FunctionPasses. Given some Function,
+/// CompilationContext, and provided Pipeline, it will run all passes on the
+/// Function. Enables easier debugging given runPrePass() and runPostPass()
+/// calls, which can be modified to run some code before or before every pass.
+class FunctionPassManager {
```

Comment:
BTW, it could useful to add a name to the FPM, similar to stageName in Swift PM (https://github.com/apple/swift/blob/master/include/swift/SILOptimizer/PassManager/PassManager.h). This is useful for debugging and logging.